### PR TITLE
IGNITE-23168 Event consumption time metric in CDC consumer

### PR DESCRIPTION
**Pull request by maksaska**
Within this commit there are 2 metrics implemented:
* EventsConsumptionTime - Time for which this CDC client processed the WAL segment events, in nanoseconds.
* EventsConsumptionTimeTotal - The total time for which this CDC client processed the WAL segments events, in nanoseconds.

It will help to monitor the throughput of the transfered cache entries as well as finish time of the proccess.

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [x] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: docs/_docs/persistence/change-data-capture.adoc [hunk 1]
```diff
@@ -125,6 +125,8 @@ The following metrics are provided by the application (additional metrics can be
 | BinaryMetaDir | Binary meta-directory the application reads data from.
 | MarshallerDir | Marshaller directory the application reads data from.
 | CdcDir | The CDC directory the application reads data from.
+| EventsConsumptionTime | Time for which this CDC client processed the WAL segment events, in nanoseconds.
```
Comment (shishkovilja) [inline docs/_docs/persistence/change-data-capture.adoc:128]:
```suggestion
| EventsConsumptionTime | Time of WAL segment events processing by a CDC consumer, in nanoseconds.
```

File: docs/_docs/persistence/change-data-capture.adoc [hunk 2]
```diff
@@ -125,6 +125,8 @@ The following metrics are provided by the application (additional metrics can be
...
+| EventsConsumptionTimeTotal | The total time for which this CDC client processed the WAL segments events, in nanoseconds.
```
Comment (shishkovilja) [inline docs/_docs/persistence/change-data-capture.adoc:129]:
```suggestion
| EventsConsumptionTimeTotal | Total time of events processing by a CDC consumer, in nanoseconds.
```

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 3]
```diff
@@ -183,6 +185,29 @@ public class CdcMain implements Runnable {
...
+
+    /** */
+    public static final String EVENTS_CONSUMPTION_TIME_DESC =
+        "Time for which this CDC client processed the WAL segment events, in nanoseconds.";
+
+    /** */
+    public static final String EVENTS_CONSUMPTION_TIME_TOTAL = "EventsConsumptionTimeTotal";
+
+    /** */
+    public static final String EVENTS_CONSUMPTION_TIME_TOTAL_DESC =
```
Comment (shishkovilja) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:208]: Fix it in the same way as the documentation.

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 4]
```diff
@@ -183,6 +185,29 @@ public class CdcMain implements Runnable {
...
+        NANOSECONDS.convert(100, MILLISECONDS),
+        NANOSECONDS.convert(250, MILLISECONDS),
+        NANOSECONDS.convert(1000, MILLISECONDS)
+    };
+
+    /** */
+    public static final String EVENTS_CONSUMPTION_TIME = "EventsConsumptionTime";
+
+    /** */
+    public static final String EVENTS_CONSUMPTION_TIME_DESC =
```
Comment (shishkovilja) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:201]: Fix it in the same way as the documentation.

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 5]
```diff
@@ -204,6 +229,12 @@ public class CdcMain implements Runnable {
     /** Time of last segment consumption. */
     private AtomicLongMetric lastSegmentConsumptionTs;
 
+    /** Metric shows duration of the last WAL segment events consumption. */
```
Comment (shishkovilja) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:232]: Fix it in the same way as the documentation.

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 6]
```diff
@@ -204,6 +229,12 @@ public class CdcMain implements Runnable {
...
+    private HistogramMetricImpl eventsConsumptionTime;
+
+    /** Metric shows the total duration of WAL segment events consumption. */
```
Comment (shishkovilja) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:235]: Fix it in the same way as the documentation.

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 7]
```diff
@@ -616,6 +652,18 @@ private void consumeSegmentActively(IgniteWalIteratorFactory.IteratorParametersB
...
+        finally {
+            updateWalProcessingMetrics(System.nanoTime() - start);
+        }
+    }
+
+    /**
+     * Updates consumer event coonsumption metrics.
+     * @param duration - Operation durtion.
+     */
+    private void updateWalProcessingMetrics(long duration) {
```
Comment (shishkovilja) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:664]: This method consisting of two lines is used only once, why should we add it?

**Review by shishkovilja — CHANGES_REQUESTED** [review]

File: docs/_docs/persistence/change-data-capture.adoc (hunk 2 above)
Comment (maksaska) [reply docs/_docs/persistence/change-data-capture.adoc]: Agree

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java (hunk 5 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java]: Fixed

File: docs/_docs/persistence/change-data-capture.adoc (hunk 1 above)
Comment (maksaska) [reply docs/_docs/persistence/change-data-capture.adoc]: Agree

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java (hunk 3 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:208]: Fixed it

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java (hunk 6 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java (hunk 7 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:664]: I added it for consistency between `eventsConsumptionTime` and `eventsConsumptionTimeTotal`. By calling this method we fix the duration as a method variable, and assign the same value to both metrics.

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java (hunk 4 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:201]: Done!

File: modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java [hunk 8]
```diff
@@ -269,6 +261,55 @@ private void checkMetrics(long expCnt, Function<String, Long> longMetric, Functi
...
+    private void checkWalProcessingMetrics(Function<String, Long> longMetric, Function<String, long[]> longMetricArray) {
+        assertNotNull(longMetric.apply(EVENTS_CONSUMPTION_TIME_TOTAL));
+
+        assertTrue(longMetric.apply(EVENTS_CONSUMPTION_TIME_TOTAL) > 0);
+
+        assertTrue(Arrays.stream(longMetricArray.apply(EVENTS_CONSUMPTION_TIME)).sum() > 0);
+    }
+
+    /** @return MBean for CDC metrics */
+    private DynamicMBean getJmxCdcReg(CdcMain cdc) {
```
Comment (shishkovilja) [inline modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java:290]:
```suggestion
    private DynamicMBean jmxRegistry(CdcMain cdc) {
```

File: modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java [hunk 9]
```diff
@@ -269,6 +261,55 @@ private void checkMetrics(long expCnt, Function<String, Long> longMetric, Functi
...
+        IgniteConfiguration cfg = getFieldValue(cdc, "igniteCfg");
+
+        return metricRegistry(cdcInstanceName(cfg.getIgniteInstanceName()), null, "cdc");
+    }
+
+    /** @return {@link MetricRegistry} */
+    private MetricRegistry getMetricRegistry(CdcMain cdc) {
```
Comment (shishkovilja) [inline modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java:297]:
```suggestion
    private MetricRegistry registry(CdcMain cdc) {
```

File: modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java [hunk 10]
```diff
@@ -269,6 +261,55 @@ private void checkMetrics(long expCnt, Function<String, Long> longMetric, Functi
...
+        return getFieldValue(cdc, "mreg");
+    }
+
+    /** */
+    private Function<String, ?> jmxVal(DynamicMBean jmxCdcReg) {
```
Comment (shishkovilja) [inline modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java:302]:
```suggestion
    private Function<String, ?> jmxValue(DynamicMBean jmxCdcReg) {
```

**Review by shishkovilja — CHANGES_REQUESTED** [review]

File: modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java (hunk 8 above)
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java]: Done!

File: modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java (hunk 9 above)
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java]: Fixed!

File: modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java (hunk 10 above)
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java]: Renamed

**Review by shishkovilja — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 11]
```diff
@@ -612,6 +629,10 @@ private void consumeSegmentActively(IgniteWalIteratorFactory.IteratorParametersB
 
             if (interrupted)
                 throw new IgniteException("Change Data Capture Application interrupted");
+
+            long duration = U.currentTimeMillis() - start;
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:633]: can be inlined

File: modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java [hunk 12]
```diff
@@ -222,37 +224,32 @@ protected GridAbsPredicate sizePredicate(
 
     /** */
     protected void checkMetrics(CdcMain cdc, int expCnt) throws Exception {
-        IgniteConfiguration cfg = getFieldValue(cdc, "igniteCfg");
```
Comment (timoninmaxim) [inline modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java:225]: Why do you need this changes?

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java (hunk 11 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java]: Agree. I fixed it

File: modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java (hunk 12 above)
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/cdc/AbstractCdcTest.java]: It's redundant code, that I left by mistake from previous commits. Fixed it

**Review by timoninmaxim — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 13]
```diff
@@ -203,6 +213,9 @@ public class CdcMain implements Runnable {
     /** Time of last segment consumption. */
     private AtomicLongMetric lastSegmentConsumptionTs;
 
+    /** Time of WAL segment events processing by a CDC consumer, in milliseconds. */
+    private HistogramMetricImpl eventsConsumptionTime;
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:217]: events -> evt

Move the declaration closer to `evtCaptureTime` and describe in javadocs the time range (see how `evtCaptureTime` javadoc is written). Do the same for metric description.

File: modules/core/src/test/java/org/apache/ignite/cdc/CdcSelfTest.java [hunk 14]
```diff
@@ -291,6 +294,15 @@ private void readAll(UserCdcConsumer cnsmr, boolean offsetCommit) throws Excepti
 
         checkMetrics(cdcMain, offsetCommit ? KEYS_CNT : ((KEYS_CNT + 3) * 2 + KEYS_CNT));
 
+        // Metric check awaits the CDC consumer to finish
+        MetricRegistry mreg = getFieldValue(cdcMain, "mreg");
```
Comment (timoninmaxim) [inline modules/core/src/test/java/org/apache/ignite/cdc/CdcSelfTest.java:298]: inline it

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java (hunk 13 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java]: Done!

File: modules/core/src/test/java/org/apache/ignite/cdc/CdcSelfTest.java (hunk 14 above)
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/cdc/CdcSelfTest.java]: Done!

File: modules/core/src/test/java/org/apache/ignite/cdc/CdcSelfTest.java [hunk 15]
```diff
@@ -291,6 +294,15 @@ private void readAll(UserCdcConsumer cnsmr, boolean offsetCommit) throws Excepti
...
+
+        long[] evtCaptureTime = mreg.<HistogramMetric>findMetric(EVENTS_CONSUMPTION_TIME).value();
+
+        assertFalse(F.isEmpty(evtCaptureTime));
+
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/cdc/CdcSelfTest.java]: Done!

File: modules/core/src/test/java/org/apache/ignite/cdc/CdcSelfTest.java [hunk 16]
```diff
@@ -291,6 +294,15 @@ private void readAll(UserCdcConsumer cnsmr, boolean offsetCommit) throws Excepti
 
         checkMetrics(cdcMain, offsetCommit ? KEYS_CNT : ((KEYS_CNT + 3) * 2 + KEYS_CNT));
 
+        // Metric check awaits the CDC consumer to finish
```
[the remark this answers is not in the document]
Comment (maksaska) [reply modules/core/src/test/java/org/apache/ignite/cdc/CdcSelfTest.java]: Done!

**Comment by maksaska** [comment]
> 1. Metric name is a little bit misleading
> 2. Metric isn't updated for passive mode
Done!

File: docs/_docs/persistence/change-data-capture.adoc [hunk 17]
```diff
@@ -125,6 +125,7 @@ The following metrics are provided by the application (additional metrics can be
 | BinaryMetaDir | Binary meta-directory the application reads data from.
 | MarshallerDir | Marshaller directory the application reads data from.
 | CdcDir | The CDC directory the application reads data from.
+| WalProcessingTime | Time of WAL segment processing, in milliseconds.
```
Comment (timoninmaxim) [inline docs/_docs/persistence/change-data-capture.adoc:128]: SegmentConsumingTime

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 18]
```diff
@@ -575,13 +585,20 @@ private boolean consumeSegment(Path segment) {
...
+            if (consumeSegmentPassively(builder)) {
+                walProcessingTime.value(U.currentTimeMillis() - start);
+
                 return true;
+            }
         }
         else
             consumeSegmentActively(builder);
 
+        walProcessingTime.value(U.currentTimeMillis() - start);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:600]: code duplication

File: docs/_docs/persistence/change-data-capture.adoc (hunk 17 above)
Comment (maksaska) [reply docs/_docs/persistence/change-data-capture.adoc]: Done!

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java (hunk 18 above)
Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java]: Removed code duplication

File: modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java [hunk 19]
```diff
@@ -212,6 +215,9 @@ public class CdcMain implements Runnable {
      */
     private HistogramMetricImpl evtCaptureTime;
 
+    /** Metric represents time between creating {@link WALIterator} and processing it, in milliseconds. */
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java:218]: prossing -> finish consuming it

Comment (maksaska) [reply modules/core/src/main/java/org/apache/ignite/internal/cdc/CdcMain.java]: Done!

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**